csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.ql

```diff
@@ -0,0 +1,98 @@
+/**
+ * @name Use of a hash function without a salt
+ * @description Hashed passwords without a salt are vulnerable to dictionary attacks.
+ * @kind path-problem
+ * @id cs/hash-without-salt
+ * @tags security
+ *       external/cwe-759
+ */
+
+import csharp
+import semmle.code.csharp.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/** The C# class `System.Security.Cryptography.SHA...` other than the weak `SHA1`. */
+class SHA extends RefType {
+  SHA() { this.getQualifiedName().regexpMatch("System\\.Security\\.Cryptography\\.SHA\\d{2,3}") }
+}
+
+class HashAlgorithmProvider extends RefType {
+  HashAlgorithmProvider() {
+    this.hasQualifiedName("Windows.Security.Cryptography.Core", "HashAlgorithmProvider")
+  }
+}
+
+/** The method call `ComputeHash()` declared in `System.Security.Cryptography.SHA...`. */
+class ComputeHashMethodCall extends MethodCall {
+  ComputeHashMethodCall() {
+    this.getQualifier().getType() instanceof SHA and
+    this.getTarget().hasName("ComputeHash")
+  }
+}
+
+/** The method call `ComputeHash()` declared in `System.Security.Cryptography.SHA...`. */
+class HashDataMethodCall extends MethodCall {
+  HashDataMethodCall() {
+    this.getQualifier().getType() instanceof HashAlgorithmProvider and
+    this.getTarget().hasName("HashData")
+  }
+}
```

Comment:
I would restructure this a bit:
```ql
class HashMethod extends Method {
  HashMethod() {
    this.getDeclaringType() instanceof SHA and
    this.hasName("ComputeHash")
    or
    this.getDeclaringType() instanceof HashAlgorithmProvider and
    this.hasName("HashData")
  }
}

...

  override predicate isSink(DataFlow::Node sink) {
    exists(MethodCall mc |
      sink.asExpr() = mc.getArgument(0) and
      mc.getTarget() instanceof HashMethod
    )
  }
```

Reply:
Thanks @hvitved for reviewing this PR. I've made requested changes.